metadata-ingestion/src/datahub/ingestion/source/sql_queries.py

```diff
@@ -194,20 +336,158 @@ def get_workunits_internal(
     ) -> Iterable[Union[MetadataWorkUnit, MetadataChangeProposalWrapper]]:
         logger.info(f"Parsing queries from {os.path.basename(self.config.query_file)}")
 
+        if self.config.enable_streaming:
+            logger.info("Using streaming processing for optimal memory usage")
+            yield from self._process_queries_streaming()
+        else:
+            logger.info("Using batch processing (all queries loaded into memory)")
+            yield from self._process_queries_batch()
+
+        # Log processing statistics
+        self._log_processing_statistics()
+
+    def _process_queries_streaming(
+        self,
+    ) -> Iterable[Union[MetadataWorkUnit, MetadataChangeProposalWrapper]]:
+        """Process queries in streaming fashion to minimize memory usage."""
+        with self.report.new_stage("Streaming query processing"):
```

Comment:
To be considered:

In Python 3.12+: `itertools.batched`

Given we also support Python <3.12, we may use third-party libraries like [more-itertools.chunked](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.chunked), which I think we already include.